### PR TITLE
LocalBearTestHelper: Add get dependency mechanism

### PR DESCRIPTION
**Pull request by refeed**
Add mechanism to get bear dependencies in
`LocalBearTestHelper.get_results()`.

Closes https://github.com/coala/coala/issues/2860

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/testing/LocalBearTestHelper.py [hunk 1]
```diff
@@ -26,20 +26,45 @@ def execute_bear(bear, *args, **kwargs):
...
+                                                  force_linebreaks,
+                                                  create_tempfile,
+                                                  tempfile_kwargs,
+                                                  settings)
+
     with prepare_file(lines, filename,
                       force_linebreaks=force_linebreaks,
                       create_tempfile=create_tempfile,
-                      tempfile_kwargs=tempfile_kwargs) as (file, fname), \
-        execute_bear(local_bear, fname, file,
```
Comment (jayvdb) [inline coalib/testing/LocalBearTestHelper.py]: You should be able to do:

```diff
-execute_bear(local_bear, fname, file,
+execute_bear(local_bear, fname, file, dependency_results=dep_results,
```

And initialise `if not local_bear.BEAR_DEPS: dep_results = None`.

If that doesnt work, IMO `Bear` needs to be enhanced to detect kwarg `dependency_results` and re-route that argument so it works correctly.

File: coalib/testing/LocalBearTestHelper.py [hunk 2]
```diff
@@ -26,20 +26,45 @@ def execute_bear(bear, *args, **kwargs):
     return list(bear_output_generator)
 
 
+def get_local_bear_run_settings_dict(local_bear, settings_dict):
```
Comment (jayvdb) [inline coalib/testing/LocalBearTestHelper.py]: This should not be necessary here.

It should be a method of `Bear` or `FunctionMetadata`, if something similar & sensible doesnt already exist.

Also this algorithm should be recursive (c.f. https://github.com/coala/coala/pull/3663)

File: coalib/testing/LocalBearTestHelper.py (hunk 1 above)
Comment (refeed) [inline coalib/testing/LocalBearTestHelper.py]: Well, when I try that, I got this:
```
E           AssertionError: Bear returned None on execution
E            
E           Running bear TestDepBearA...
E           Bear TestDepBearA failed to run on file /tmp/tmpuf5mq4w6. Take a look at debug messages (`-V`) for further information.
E           The bear TestDepBearA raised an exception. If you are the author of this bear, please make sure to catch all exceptions. If not and this error annoys you, you might want to get in contact with the author of this bear.
E           
E           Traceback information is provided below:
E           
E           Traceback (most recent call last):
E             File "/home/rafid/GitHub/coala/coalib/bears/Bear.py", line 292, in execute
E               result = self.run_bear_from_section(args, kwargs)
E             File "/home/rafid/GitHub/coala/coalib/bears/Bear.py", line 285, in run_bear_from_section
E               return self.run(*args, **kwargs)
E           TypeError: run() got an unexpected keyword argument 'dependency_results'
```

That's caused by run method of bear which doesn't have dependency doesn't have `dependency_results` arg.

> If that doesnt work, IMO Bear needs to be enhanced to detect kwarg dependency_results and re-route that argument so it works correctly.
Do we need to create an issue for this?

File: coalib/testing/LocalBearTestHelper.py (hunk 2 above)
Comment (refeed) [inline coalib/testing/LocalBearTestHelper.py]: Right, I found a correct method in `FunctionMetadata` called `filter_parameters`

File: coalib/testing/LocalBearTestHelper.py (hunk 1 above)
Comment (jayvdb) [inline coalib/testing/LocalBearTestHelper.py]: No, it doesnt need a separate issue. It does need to be a separate commit.

Comment (refeed) [inline coalib/testing/LocalBearTestHelper.py]: ok

**Comment by refeed** [comment]
@jayvdb updated!

File: coalib/testing/LocalBearTestHelper.py [hunk 3]
```diff
@@ -33,12 +33,28 @@ def get_results(local_bear,
                 create_tempfile=True,
                 tempfile_kwargs={},
                 settings={}):
+    deps_results = dict()
```
Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: you can move that below into the `if`, as in the `else` clause you set it to `None` anyway^^

Comment (refeed) [inline coalib/testing/LocalBearTestHelper.py]: right

**Comment by refeed** [comment]
updated!

File: tests/coalaJSONTest.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def test_show_all_bears(self):
                 coala.main, 'coala', '--json', '-B', '-I')
             self.assertEqual(retval, 0)
             output = json.loads(stdout)
-            self.assertEqual(len(output['bears']), 8)
+            self.assertEqual(len(output['bears']), 13)
```
Comment (jayvdb) [inline tests/coalaJSONTest.py]: Eh, why did this grow?  Is this because of your changes in `LocalBearTestHelper.py`, or caused by your other commit which changes `Bear.py`?

I can only see 8 in `tests/test_bears/` and the extra ones are in `TestBearDep`.  but why are they only appearing now, and not before?  Were they hidden intentionally?

I think it would be good to assert the exact list, as that helps show there are multiple levels of dependencies, and different types (some are hidden in the same file).

File: coalib/testing/LocalBearTestHelper.py [hunk 5]
```diff
@@ -33,12 +33,28 @@ def get_results(local_bear,
                 create_tempfile=True,
                 tempfile_kwargs={},
                 settings={}):
+    if local_bear.BEAR_DEPS:
+        # Get results of bear's dependencies first
+        deps_results = dict()
+        for bear in local_bear.BEAR_DEPS:
```
Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: what if the bear-dependency has itself dependencies?

File: tests/collecting/CollectorsTest.py [hunk 6]
```diff
@@ -317,7 +317,12 @@ def test_get_all_bears(self):
                  'SpaceConsistencyTestBear',
                  'TestBear',
                  'ErrorTestBear',
-                 'RaiseTestBear'})
+                 'RaiseTestBear',
+                 'TestDepBearA',
+                 'TestDepBearBDependsA',
+                 'TestDepBearCDependsB',
+                 'TestDepBearAA',
+                 'TestDepBearDependsAAndAA'})
```
Comment (Makman2) [inline tests/collecting/CollectorsTest.py]: what has changed that `Collectors` is affected? :O

Comment (Makman2) [inline tests/collecting/CollectorsTest.py]: nvm

/me should scroll down before when he wants to posts a comment :tada:

File: coalib/testing/LocalBearTestHelper.py (hunk 5 above)
Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: nvm it's recursive

File: tests/collecting/CollectorsTest.py (hunk 6 above)
Comment (jayvdb) [inline tests/collecting/CollectorsTest.py]: ah, that was my problem too!

File: tests/coalaJSONTest.py (hunk 4 above)
Comment (jayvdb) [inline tests/coalaJSONTest.py]: Err, ignore this.

File: tests/testing/LocalBearTestHelperTest.py [hunk 7]
```diff
@@ -2,6 +2,9 @@
 import unittest
 
 from tests.test_bears.TestBear import TestBear
+from tests.test_bears.TestBearDep import TestDepBearBDependsA,\
+                                         TestDepBearCDependsB, \
+                                         TestDepBearDependsAAndAA
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: use braces instead of backslashes ;)

**Comment by Makman2** [comment]
Btw: this won't be necessary any more with the next-gen core ;) dependency results will be stored at `self.dependency_results` inside a bear :+1:

File: coalib/bears/Bear.py [hunk 8]
```diff
@@ -288,6 +288,14 @@ def execute(self, *args, debug=False, **kwargs):
         name = self.name
         try:
             self.debug('Running bear {}...'.format(name))
+
+            # If `dependency_results` kwargs is defined but there's no
+            # dependency results (usually in Bear that has no dependency)
+            # delete the `dependency_results` kwargs, since most Bears doesn't
```
Comment (Makman2) [inline coalib/bears/Bear.py]: doesn't --> don't

(same in the commit message ;D)

File: coalib/bears/Bear.py [hunk 9]
```diff
@@ -288,6 +288,14 @@ def execute(self, *args, debug=False, **kwargs):
...
+            # define `dependency_results` kwargs in its `run()` argument.
+            if kwargs.get('dependency_results', False) is None:
```
Comment (Makman2) [inline coalib/bears/Bear.py]: Hmm this optional value fallback is quite a tricky piece of code ;)

What about rephrasing to
```python
if 'dependency_results' in kwargs and kwargs['dependency_results'] is None:
    del ...
```
That's a bit more obvious to grasp (imo)

Up to you :+1:

File: coalib/bears/Bear.py [hunk 10]
```diff
@@ -288,6 +288,14 @@ def execute(self, *args, debug=False, **kwargs):
         name = self.name
         try:
             self.debug('Running bear {}...'.format(name))
+
+            # If `dependency_results` kwargs is defined but there's no
```
Comment (Makman2) [inline coalib/bears/Bear.py]: there's -> there are

(same in commit message)

File: coalib/bears/Bear.py [hunk 11]
```diff
@@ -288,6 +288,14 @@ def execute(self, *args, debug=False, **kwargs):
...
+            # dependency results (usually in Bear that has no dependency)
+            # delete the `dependency_results` kwargs, since most Bears doesn't
+            # define `dependency_results` kwargs in its `run()` argument.
```
Comment (Makman2) [inline coalib/bears/Bear.py]: argument --> function

(same in commit message)

**Comment by Makman2** [comment]
For `Bear: Check dependency_results kwargs in execute()` (69a22d8): Is the `Closes XXX` reference correct? I think this is only one step to closing the issue. You should place a `Related to XXX` reference instead :+1:

**Comment by Makman2** [comment]
What is 69a22d8 needed for actually? Do some errors occur otherwise inside `LocalBearTestHelper`?

**Comment by jayvdb** [comment]
@makman2 : See https://github.com/coala/coala/pull/4399#discussion_r125161246

File: coalib/bears/Bear.py (hunk 9 above)
Comment (refeed) [inline coalib/bears/Bear.py]: ok

File: tests/collecting/CollectorsTest.py (hunk 6 above)
Comment (refeed) [inline tests/collecting/CollectorsTest.py]: :)

File: coalib/bears/Bear.py (hunk 9 above)
Comment (refeed) [inline coalib/bears/Bear.py]: I'll take that

File: coalib/bears/Bear.py (hunk 10 above)
Comment (refeed) [inline coalib/bears/Bear.py]: right

**Comment by refeed** [comment]
updated!

**Comment by jayvdb** [comment]
ack 66972ed

**Comment by jayvdb** [comment]
reack d232984 633f6e5

**Comment by jayvdb** [comment]
@rultor merge

**Comment by jayvdb** [comment]
@rultor merge

**Comment by Makman2** [comment]
@rultor merge
